### ISSUE TITLE
The static DefaultTemplate is not used for inner classes in codebehind for TypedJSON

### ISSUE THREAD

**Issue by chrhol** [Starcounter/Starcounter#3373]
The following code-snippet contains both the working (using the template for the root object) and non-working (using the default template for the child):

**SomeJson.json**
```javascript
{
    "InnerObject": {
        "InnerProperty": ""
    }
}
```

**SomeJson.json.cs**
```csharp
[SomeJson_json]
public partial class SomeJson : Json {
    [SomeJson_json.InnerObject]
    public partial class InnerObjectJson : Json {
         static InnerObjectJson() {
              // Does not work, the DefaultTemplate for InnerObjectJson is not used.
              InnerObjectJson.DefaultTemplate.InnerProperty.Bind = "DoesNotWorkHere";

              // This works however:
              SomeJson.DefaultTemplate.InnerObject.InnerProperty = "ThisOneIsUsed".
         }
    }
}
```

Both lines in the static constructor should yield the same behaviour. 

Note: This is only a problem for objects, arrays works as expected.

**Comment by chrhol** [comment]
Found by @miyconst

**Comment by chrhol** [comment]
I might have been a bit too quick in making this assumption. The inner object cannot use the default template since you can have more than one instance of the outer template

Consider this example: 
```csharp
SomeJsonTemplate template1 = new SomeJsonTemplate();
template1.InnerObject.InnerProperty.Bind = "Binding";
SomeJsonTemplate template2 = new SomeJsonTemplate();
template2.InnerObject.InnerProperty.Bind = "CompletelyDifferentBinding";

SomeJson json1 = new SomeJson() { Template = template1; }
SomeJson json2 = new SomeJson() { Template = template2; }
```

Here you will have two instances of the same template, and the template for `InnerObject` is, and should be, two completely different instances. So the way to access inner templates is throught the parent.

I need to think about this some more, maybe inner objects should not have the `DefaultTemplate` property at all.

**Comment by miyconst** [comment]
Hm, does this mean that I can change the "bind" property per instance and not per class?

**Comment by chrhol** [comment]
> Hm, does this mean that I can change the "bind" property per instance and not per class?

It gets a little confusing. You can change it per instance of the template (or class if you will) but not per instance of Json. Technically you could have one instance of the template for one instance of Json, however this will not be performant at all. 

We haven't really promoted this, we aways use the default template, but you *can* if you like do as the above to have separate templates that are a bit different. In most cases having separate templates should not be needed but might in some cases.

**Comment by miyconst** [comment]
I am still puzzled. Any JSON can be attached to any other JSON.

So, if we have `A`, `B`, `C` classes, then I can set template values by:

```cs
A.DefaultTemplate.Property.Bind = "PropertyBind";
B.DefaultTemplate.Property.Bind = "PropertyBind";
C.DefaultTemplate.Property.Bind = "PropertyBind";
``` 

Now, if `B` belongs to `A` I have to change it:

```cs
A.DefaultTemplate.B.Property.Bind = "PropertyBind";
```

Now, what will happen if I attach `A` as a property of `C`? I have to reset all the template settings using `C.DefaultTemplate`?

**Comment by chrhol** [comment]
Yes, I can understand that. I feel a bit confused trying to find some good way to explain it as well :)

Maybe part of it's the `DefaultTemplate` that sometimes should be used and sometime not. I will get back to you, need to be think about this a bit.

**Comment by chrhol** [comment]
> Now, what will happen if I attach A as a property of C? I have to reset all the template settings using C.DefaultTemplate?

I started writing down an answer, but I'm not sure it makes sense. The `DefaultTemplate` messes it up. I think we need to change so that `DefaultTemplate` is not public, or maybe removed (doesn't really matter if we hide it then it can be changed whichever way). It should not be used by applicationcode, since it is not always correct to use it.

I guess it depends on how A, B and C are declared. Are they all separate TypedJson they have nothing with each other to do. And the template for the *property* in A that you assign C to is not C.DefaultTemplate (and there is no way it can be).

**Comment by chrhol** [comment]
We could implement a method in the codebehind, that will be called when the template is instantiated. The template itself is sent in as parameter. This will allow setting metadata during creation (like bindings) of the template. This method would also be called for all inner classes if it is declared. Also we are in control of when it's called instead of depending on the static constructor, which makes it easier to change things underneath. 

**SomeJson.json**
```javascript
{
  "Header": "",
  "InnerObject": {
    "InnerProperty": ""
  }
}
```

**SomeJson.json.cs**
```csharp
partial class SomeJson : Json {
  private static void Init(SomeJsonTemplate template) {
    template.Header.Bind = "Binding";
  }

  [SomeJson_json.InnerObject]
  partial class InnerObjectJson : Json {
    private static void Init(InnerObjectTemplate template) {
      template.InnerProperty.Bind = "Binding";
    }
  }
}
```

This way, you always have the correct template. You could of course set the binding for `InnerProperty` from the parent if you like: `template.InnerObject.InnerProperty.Bind = "Binding"` and be certain that it is the correct template for the inner TypedJson you are modifying.

This requires some modification and enhancments in the codebehind-parser though, so it's probably dependent on #3439

**Comment by miyconst** [comment]
I really like your idea with the `Init` function!

- We rid of `DefaultTemplate` and static constructor.
- Able to set different settings per each instance.
- Don't have to keep initialization code in the `OnData` method and then set `this.Data = null` to invoke it.

**Comment by chrhol** [comment]
Part of #3727 

Will be solved by having a specific init method for templates.
